Error bindings

spacebear21 opened the pull request:
Addresses #48

This exposes new top-level error types, and in cases where inner types are transparent also exposes those. Opaque inner types are kept as simple String representations.

PayjoinError is obsoleted by the new error types and removed in the last commit.

DanGould commented on src/receive/error.rs:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+impl From<payjoin::receive::ReplyableError> for ReplyableError {
+    fn from(value: payjoin::receive::ReplyableError) -> Self {
+        use payjoin::receive::ReplyableError::*;
+
+        match value {
+            Payload(_) => ReplyableError::Payload { message: value.to_string() },
+            V1(_) => ReplyableError::V1 { message: value.to_string() },
+            Implementation(_) => {
+                ReplyableError::Implementation(ImplementationError { message: value.to_string() })
+            }
+        }
+    }
+}
```
By using `value.to_string()` instead of the variant `e.to_string`, isn't the thiserror annotation prefix redundant?

e.g. won't 
```rs
    #[error("Error while validating V1 request: {message}")]
    V1 { message: String },`
``` 

double print "Error while validating V1 request" ?

I prefer using `value` instead of e.g. `V1(e)`'s `e` because we can remove this redundancy, but we do have to make sure we're not being redundant.

spacebear21 replied:
Hmm good point about the double printing. I can't get tests to run because of the ohttp changes so can't verify easily but it seems likely that it does double print. The solution may be to use `#[error(transparent)]` everywhere...

DanGould replied:
The OHTTP changes should work with production servers now. Which problem are you experiencing?

spacebear21 replied:
The problem is that I need to upgrade the `payjoin-directory` dev dependency, but if I upgrade it to say:

```toml
payjoin-directory = { git = "https://github.com/payjoin/rust-payjoin", branch = "bindings-0.23", features = ["_danger-local-https"] }
```

I get:

```
❯ cargo test --features=_danger-local-https
    Updating git repository `https://github.com/payjoin/rust-payjoin.git`
    Updating crates.io index
error: failed to select a version for `payjoin`.
    ... required by package `payjoin-directory v0.0.1 (https://github.com/payjoin/rust-payjoin.git?branch=bindings-0.23#0fabc7aa)`
    ... which satisfies git dependency `payjoin-directory` of package `payjoin_ffi v0.22.1 (/Users/spacebear/Projects/payjoin-ffi)`
versions that meet the requirements `^0.22.0` are: 0.22.0

the package `payjoin-directory` depends on `payjoin`, with features: `directory` but `payjoin` does not have these features.
```

DanGould commented on src/receive/error.rs:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+impl From<payjoin::bitcoin::address::ParseError> for Error {
+    fn from(value: payjoin::bitcoin::address::ParseError) -> Self {
+        Error::V2 { message: value.to_string() }
+    }
+}
```
This can be removed by accepting a `bitcoin_ffi::Address` rather than `String` and `Network` in `Receiver::new`

spacebear21 replied:
This would be a nice follow-up, though I couldn't get it to work in `receive/uni.rs` due to:
```
the trait `Lift<UniFfiTag>` is not implemented for `bitcoin_ffi::Address`
```

DanGould replied:
I believe It just needs to be passed as `Arc<crate::bitcoin_ffi::Address>` in the argument. follow on would work.

DanGould commented on src/receive/error.rs:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+impl From<IntoUrlError> for Error {
+    fn from(value: IntoUrlError) -> Self {
+        Error::V2 { message: value.to_string() }
+    }
+}
```
All of these `From` implementations end up in the public API and I wonder if they're being used because they are indeed universal or because they make implementation less noisy.

DanGould commented on src/receive/mod.rs:
```diff
@@ -61,13 +66,10 @@ impl Receiver {
         match self.0.clone().extract_req(ohttp_relay) {
             Ok((req, ctx)) => Ok((req.into(), ctx.into())),
-            Err(e) => Err(PayjoinError::V2Error { message: e.to_string() }),
+            Err(e) => Err(Error::V2 { message: e.to_string() }),
```
This seems to be the only conversion that is not like the others. Why not an `e.into()`

`impl From<payjoin::receive::Error> for Error {`

where the appropriate variant is always chosen in the conversion? Does it conflict with something else?

spacebear21 replied:
Just an omission on my part, fixed in rebase

DanGould commented on src/receive/error.rs:
```diff
@@ -0,0 +1,135 @@
+use payjoin::IntoUrlError;
```
use `payjoin::receive` would prevent some `payjoin::receive::<errortype>` redundancy quite a few times in this file.

spacebear21 replied:
I opted for mimicking the [bdk-ffi](https://github.com/bitcoindevkit/bdk-ffi/blob/master/bdk-ffi/src/error.rs) approach of aliasing e.g. `use payjoin::receive::Error as PdkError`

DanGould commented on src/receive/error.rs:
```diff
@@ -0,0 +1,135 @@
+        match value {
+            payjoin::receive::Error::ReplyToSender(e) => Error::ReplyToSender(e.into()),
+            payjoin::receive::Error::V2(_) => Error::V2 { message: value.to_string() },
+            _ => panic!("An unexpected error occurred"),
```
A new `UnexpectedError` variant seems more appropriate than a panic here, then at least the error could be logged.

DanGould commented on src/receive/error.rs:
```diff
@@ -0,0 +1,135 @@
+    /// Error arising due to the specific receiver implementation
+    ///
+    /// e.g. database errors, network failures, wallet errors
+    #[error("{0}")]
```
I think [`#[error(transparent)]`](https://docs.rs/thiserror/latest/thiserror/#derives) was made for this use case

DanGould reviewed (commented):
I left quite a few comments on the "Expose receiver error types" commit. You may see that of the conversions may be able to be simplified.

I wonder if all of the `impl From` blocks are appropriate on the public API. Most of them certainly are, but some from specific variants make me wonder.

I wonder if we're complicating our life by using these #[error(message)] macros or if they're relatively self documented. Did you think they were just better documenting than comments, or something else? We don't have them on the error types upstream, and the bound errors will print different messages because of the #[error(...)] macros

Over all I think this constitutes a much more workable experience for users, though we're going to need to test it, and I can see many places where our API would benefit users to upgrade to using bitcoin_ffi types instead of the raw e.g. String or u64 types.

DanGould commented on src/send/mod.rs:
```diff
@@ -2,11 +2,14 @@ use std::io::Cursor;
 use std::sync::{Arc, Mutex};
 
-use crate::error::PayjoinError;
+use error::{BuildSenderError, CreateRequestError, EncapsulationError, ResponseError};
```
In [rust-payjoin we pub use Error types](file:///Users/dan/f/dev/payjoin/target/doc/payjoin/send/index.html#enums) so that they can be found directly in their module. I think we ought to do that here, for appropriate modules as well.

DanGould commented on src/receive/mod.rs:
```diff
@@ -441,7 +438,7 @@ impl PayjoinProposal {
 //     use super::*;
 
-//     fn get_proposal_from_test_vector() -> Result<UncheckedProposal, PayjoinError> {
+//     fn get_proposal_from_test_vector() -> Result<UncheckedProposal, Error> {
```
he thorough

DanGould reviewed (approved):
The one major concern was [addressed](https://github.com/LtbLightning/payjoin-ffi/pull/58#discussion_r2011015094)

Everything else mostly relates to form rather than function, and given our time constraints, I'm inclined to move forward.

I *do* think an audit of thiserror macro usage to impl Display is in order, but that's secondary to getting the semantic error handling correct. And this gets a whole lot closer.

ACK f1ffa2fbbde0cf4dbed01ae9c32b5b07521fb784

DanGould commented:
something was merged that requires rebase before this goes in

DanGould commented:
Sounds also like `message` field on error types needs to be renamed `msg` as well 

see: https://github.com/LtbLightning/payjoin-ffi/pull/58

spacebear21 commented:
Resolved conflicts, renamed `message` fields on all new error types to `msg`, and removed the error aliases in favor of `use payjoin::receive`/`use payjoin::send`.

Follow-up todos:
- Remove fully qualified names in the `mod.rs` files as well. 
- Evaluate using `#[error(transparent)]` on all error types instead of potentially double-printing messages.
- Accept a `bitcoin_ffi::Address` rather than String and Network in `Receiver::new`
- Audit From implementations that will end up in the public API

DanGould commented on src/receive/error.rs:
```diff
@@ -0,0 +1,142 @@
+    msg: String,
+}
+
+impl From<receive::Error> for Error {
+    fn from(value: receive::Error) -> Self {
+        match value {
+            receive::Error::ReplyToSender(e) => Error::ReplyToSender(e.into()),
+            receive::Error::V2(_) => Error::V2 { msg: value.to_string() },
+            _ => Error::Unexpected,
+        }
+    }
+}
+
+impl From<receive::ReplyableError> for ReplyableError {
+    fn from(value: receive::ReplyableError) -> Self {
+        match value {
+            receive::ReplyableError::Payload(_) => {
+                ReplyableError::Payload { msg: value.to_string() }
+            }
+            receive::ReplyableError::V1(_) => ReplyableError::V1 { msg: value.to_string() },
+            receive::ReplyableError::Implementation(_) => {
+                ReplyableError::Implementation(Arc::new(ImplementationError {
+                    msg: value.to_string(),
+                }))
+            }
+        }
+    }
+}
+
+impl From<payjoin::bitcoin::address::ParseError> for Error {
+    fn from(value: payjoin::bitcoin::address::ParseError) -> Self {
+        Error::V2 { msg: value.to_string() }
+    }
+}
+
+impl From<IntoUrlError> for Error {
+    fn from(value: IntoUrlError) -> Self {
+        Error::V2 { msg: value.to_string() }
+    }
+}
```
These impls definitely spook me and must be audited.

DanGould commented on src/uri/error.rs:
```diff
@@ -0,0 +1,38 @@
+    msg: String,
+}
+
+impl From<String> for PjParseError {
+    fn from(msg: String) -> Self {
+        PjParseError { msg }
+    }
+}
```
This is very smelly, address it in the audit. If we need this because we depend on bitcoin_uri errors in the public API, maybe we should reconsider those rather than making arbitrary string conversions part of the API.

Manual map_err seems a lot more appropriate in this circumstance.

DanGould reviewed (approved):
LGMT. Huge progress to have contextual bindings. Can't wait to see the implementations improve because of these.

I hope to see the conversion audit sooner than later. Letting it through is known tech debt.

84436d329d49e47c8eff997f6d38d1f7122908f3
